core/lib/dal/src/storage_logs_dal.rs

```diff
@@ -300,6 +300,51 @@ impl StorageLogsDal<'_, '_> {
         row.count > 0
     }
 
+    /// Returns addresses and the corresponding deployment miniblock numbers among the specified contract
+    /// `addresses`. `at_miniblock` allows filtering deployment by miniblocks.
+    pub async fn filter_deployed_contracts(
+        &mut self,
+        addresses: impl Iterator<Item = Address>,
+        at_miniblock: Option<MiniblockNumber>,
+    ) -> sqlx::Result<HashMap<Address, MiniblockNumber>> {
+        let (bytecode_hashed_keys, address_by_hashed_key): (Vec<_>, HashMap<_, _>) = addresses
+            .map(|address| {
+                let hashed_key = get_code_key(&address).hashed_key().0;
+                (hashed_key, (hashed_key, address))
+            })
+            .unzip();
+        let max_miniblock_number = at_miniblock.map_or(u32::MAX, |number| number.0);
+        let rows = sqlx::query!(
+            r#"
+            SELECT DISTINCT
+                ON (hashed_key) hashed_key,
```

Comment:
I think `SELECT DISTINCT ON` can be used in other queries as well; I plan to work on using this approach separately. cc @Deniallugo; we've recently discussed a query that could be seemingly made more efficient using this approach.

Reply:
yes and thanks for creating the [ticket](https://linear.app/matterlabs/issue/PLA-803/use-select-distinct-on-queries)